### PR TITLE
Archon Hero Tree

### PR DESCRIPTION
**Pull request by Liavre**
### Description

As said before in the Oracle PR, this was split off to make that review shorter.

### Testing

- Test report URL: `/report/...`
- Screenshot(s):

### CODE REVIEW

**Comment by Liavre** [comment]
This is the main log I used to validate. (also found on the oracle PR)
https://www.warcraftlogs.com/reports/WT19GKp2VHqLarbD/#fight=19&source=122&type=healing

File: src/analysis/retail/priest/holy/CombatLogParser.tsx [hunk 1]
```diff
@@ -141,6 +141,15 @@ class CombatLogParser extends CoreCombatLogParser {
...
+    // Archon Hero Talent Display
+    PerfectedFormHoly: Talents.Archon.PerfectedFormHoly,
+    ResonantEnergyHoly: Talents.Archon.ResonantEnergyHoly,
+    ManifestedPowerHoly: Talents.Archon.ManifestedPowerHoly,
+    EmpoweredSurgesHoly: Talents.Archon.EmpoweredSurgesHoly,
+    EnergyCycleHoly: Talents.Archon.EnergyCycleHoly,
+    EnergyCompressionHoly: Talents.Archon.EnergyCompressionHoly,
+    PowerSurgeAndDivineHaloHoly: Talents.Archon.PowerSurgeAndDivineHaloHoly,
+
     //Oracle Hero Talents
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/CombatLogParser.tsx]:
```suggestion
    // Oracle Hero Talents
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+import TalentSpellText from 'parser/ui/TalentSpellText';
+import { buffedBySurgeOfLight, getHealFromSurge } from '../../../normalizers/CastLinkNormalizer';
+
+import { TALENTS_PRIEST } from 'common/TALENTS';
+import Events, { RemoveBuffEvent, RemoveBuffStackEvent } from 'parser/core/Events';
+import SPELLS from 'common/SPELLS';
+import { calculateEffectiveHealing } from 'parser/core/EventCalculateLib';
+
+//https://www.warcraftlogs.com/reports/WT19GKp2VHqLarbD#fight=19``&type=auras&source=122
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx]: either explain why this comment is here as a part of the comment or remove it

File: src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  protected combatants!: Combatants;
+
+  empoweredSurgesHealing = 0;
+
+  constructor(options: Options) {
+    super(options);
+
+    this.active = this.selectedCombatant.hasTalent(TALENTS_PRIEST.EMPOWERED_SURGES_TALENT);
+    //tracks spending of Surge of Light
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx]:
```suggestion
```
This comment doesn't add much value as the code explains itself

File: src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    );
+  }
+
+  onSurgeOfLightHeal(event: RemoveBuffEvent | RemoveBuffStackEvent) {
+    // linked heal event exists from surge of light consumption
+    const healEvent = getHealFromSurge(event);
+
+    if (healEvent) {
+      if (buffedBySurgeOfLight(event)) {
+        // calculate effective healing from bonus
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    );
+
+    this.addEventListener(
+      Events.removebuffstack.by(SELECTED_PLAYER).spell(SPELLS.SURGE_OF_LIGHT_BUFF),
+      this.onSurgeOfLightHeal,
+    );
+  }
+
+  onSurgeOfLightHeal(event: RemoveBuffEvent | RemoveBuffStackEvent) {
+    // linked heal event exists from surge of light consumption
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EmpoweredSurgesHoly.tsx]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+import TalentSpellText from 'parser/ui/TalentSpellText';
+import ItemPercentHealingDone from 'parser/ui/ItemPercentHealingDone';
+import ItemPercentDamageDone from 'parser/ui/ItemPercentDamageDone';
+import { TALENTS_PRIEST } from 'common/TALENTS';
+import Events, { DamageEvent, HealEvent } from 'parser/core/Events';
+import { calculateEffectiveDamage, calculateEffectiveHealing } from 'parser/core/EventCalculateLib';
+import SPELLS from 'common/SPELLS';
+
+//https://www.warcraftlogs.com/reports/WT19GKp2VHqLarbD#fight=19``&type=auras&source=122
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx]:
```suggestion
```
ditto: explain or remove

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  handleHaloHealing(event: HealEvent) {
+    this.totalEnergyCompressionHealing += calculateEffectiveHealing(event, ENERGY_COMPRESSION_AMP);
+  }
+
+  handleHaloDamage(event: DamageEvent) {
+    this.totalArchonHaloDamage += calculateEffectiveDamage(event, ENERGY_COMPRESSION_AMP);
+  }
+
+  getEnergyCompressionDamage() {
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx]: we try to use getters in this repo
```suggestion
  get energyCompressionDamage() {
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+  statistic() {
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.OPTIONAL(99)}
+        size="flexible"
+        category={STATISTIC_CATEGORY.HERO_TALENTS}
+        tooltip={
+          <>
+            <small>
+              This is number is included in the Divine Halo/Power Surge results already.
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx]:
```suggestion
              This number is included in the Divine Halo/Power Surge results already.
```

Optionally you can also use `SpellLink` to make this text more flavorful

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            </small>
+          </>
+        }
+      >
+        <TalentSpellText talent={TALENTS_PRIEST.ENERGY_COMPRESSION_TALENT}>
+          <div>
+            <ItemPercentHealingDone amount={this.totalEnergyCompressionHealing} />
+          </div>
+          <div>
+            <ItemPercentDamageDone amount={this.getEnergyCompressionDamage()} />{' '}
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCompressionHoly.tsx]:
```suggestion
            <ItemPercentDamageDone amount={this.energyCompressionDamage} />{' '}
```

use getter

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import BoringSpellValueText from 'parser/ui/BoringSpellValueText';
+
+import { TALENTS_PRIEST } from 'common/TALENTS';
+import SpellUsable from 'parser/shared/modules/SpellUsable';
+import { TIERS } from 'game/TIERS';
+import Events, { RemoveBuffEvent, RemoveBuffStackEvent } from 'parser/core/Events';
+import { buffedBySurgeOfLight, getHealFromSurge } from '../../../normalizers/CastLinkNormalizer';
+import SPELLS from 'common/SPELLS';
+
+//https://www.warcraftlogs.com/reports/WT19GKp2VHqLarbD#fight=19``&type=auras&source=122
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+const APOTH_MULTIPIER = 4;
+const ENERGY_CYCLE_CDR = 4;
+const LIGHT_OF_THE_NAARU_REDUCTION_PER_RANK = 0.1;
+const TWW_TIER1_2PC_CDR = 0.1;
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]: Other specs put all of their constants into a single file, might be something to consider to make it more manageable

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+  private baseHolyWordCDR = 1;
+  private modHolyWordCDR = 1;
+  private apothBuffActive = false;
+
+  constructor(options: Options) {
+    super(options);
+
+    this.active = this.selectedCombatant.hasTalent(TALENTS_PRIEST.ENERGY_CYCLE_TALENT);
+    // these two if statements get the scaling CDR for holy word reduction
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      Events.applybuff.by(SELECTED_PLAYER).spell(TALENTS_PRIEST.APOTHEOSIS_TALENT),
+      this.applyApoth,
+    );
+
+    this.addEventListener(
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS_PRIEST.APOTHEOSIS_TALENT),
+      this.removeApoth,
+    );
+
+    //tracks spending of Surge of Light
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx:67]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+      this.baseHolyWordCDR =
+        this.selectedCombatant.getTalentRank(TALENTS_PRIEST.LIGHT_OF_THE_NAARU_TALENT) *
+          LIGHT_OF_THE_NAARU_REDUCTION_PER_RANK +
+        1;
+    }
+    if (this.selectedCombatant.has2PieceByTier(TIERS.TWW1)) {
+      this.baseHolyWordCDR *= TWW_TIER1_2PC_CDR + 1;
+    }
+
+    //these are used to check if CDR needs to be scaled for Energy Cycle
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx:56]: This one is an example of a good comment as it isn't directly apparent from the listeners you wrote out.

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+  }
+
+  applyApoth() {
+    this.apothBuffActive = true;
+  }
+  removeApoth() {
+    this.apothBuffActive = false;
+  }
+  //ENERGY CYCLE VALUES
+  passWastedEnergyCycleCDR(): number {
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]:
```suggestion
  get passWastedEnergyCycleCDR(): number {
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.OPTIONAL(99)}
+        size="flexible"
+        category={STATISTIC_CATEGORY.HERO_TALENTS}
+      >
+        <BoringSpellValueText spell={TALENTS_PRIEST.ENERGY_CYCLE_TALENT}>
+          <div>
+            {formatNumber(this.energyCycleCDRActual)}s<small> reduced from Sanctify</small> <br />
+            {this.passWastedEnergyCycleCDR()}s<small> wasted</small> <br />
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]:
```suggestion
            {this.passWastedEnergyCycleCDR}s<small> wasted</small> <br />
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+  statistic() {
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.OPTIONAL(99)}
+        size="flexible"
+        category={STATISTIC_CATEGORY.HERO_TALENTS}
+      >
+        <BoringSpellValueText spell={TALENTS_PRIEST.ENERGY_CYCLE_TALENT}>
+          <div>
+            {formatNumber(this.energyCycleCDRActual)}s<small> reduced from Sanctify</small> <br />
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]:
```suggestion
            {formatNumber(this.energyCycleCDRActual)}s<small> reduced from <SpellLink spell={TALENTS_PRIEST.SANCTIFY_TALENT}/></small> <br />
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+  buffedBySurgeOfLight,
+  getHealFromSurge,
+  isSurgeOfLightFromHalo,
+} from '../../../normalizers/CastLinkNormalizer';
+/**
+ * **Perfected Form**
+ * Your healing done is increased by 10% while Apotheosis is active and for 20 sec after you cast Holy Word: Salvation.
+ */
+
+//https://www.warcraftlogs.com/reports/WT19GKp2VHqLarbD#fight=19``&type=auras&source=122
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx]:
```suggestion
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+
+  statistic() {
+    return (
+      <Statistic
+        position={STATISTIC_ORDER.OPTIONAL(99)}
+        size="flexible"
+        category={STATISTIC_CATEGORY.HERO_TALENTS}
+        tooltip={
+          <>
+            <SpellLink spell={PRIEST_TALENTS.SURGE_OF_LIGHT_TALENT} /> procs and usage from ALL
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx]:
```suggestion
            <SpellLink spell={PRIEST_TALENTS.SURGE_OF_LIGHT_TALENT} /> procs and usage from <b>all</b>
```

File: src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+      >
+        <BoringSpellValueText spell={TALENTS_PRIEST.MANIFESTED_POWER_TALENT}>
+          <div>
+            {this.surgeOfLightProcsGainedFromHalo}
+            <small> procs gained from Halo</small>{' '}
+          </div>
+          <div>
+            {this.surgeOfLightProcsOverwrittenByHalo}
+            <small> overwritten procs from Halo</small>
+          </div>
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/ManifestedPowerHoly.tsx]: Use `SpellLink` for the Halo's

File: src/analysis/retail/priest/holy/modules/talents/Archon/PowerSurgeAndDivineHaloHoly.tsx [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+
+    this.totalArchonHaloDamage += event.amount + (event.absorbed || 0);
+  }
+
+  newHaloCast(event: CastEvent) {
+    this.firstHalo = true;
+  }
+  removeArchonOut() {
+    this.firstHalo = false;
+  }
```
Comment (trevorm4) [inline src/analysis/retail/priest/holy/modules/talents/Archon/PowerSurgeAndDivineHaloHoly.tsx]:
```suggestion
  newHaloCast(event: CastEvent) {
    this.firstHalo = true;
  }
  
  removeArchonOut() {
    this.firstHalo = false;
  }
```

**Review by trevorm4 — CHANGES_REQUESTED** [review]

File: src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx (hunk 11 above)
Comment (Liavre) [reply src/analysis/retail/priest/holy/modules/talents/Archon/EnergyCycleHoly.tsx]: I have moved things to a hero tree specific file, I will move it to one main file when I do more comprehensive holy priest work.

**Comment by Liavre** [comment]
I have gone through and done the requested changes, also applying similar critiques to my Oracle module that I missed earlier today.

**Review by trevorm4 — APPROVED** [review]

**Pull request merged**